### DISCUSSION TITLE
Correct way to use IdentifiedReference with nullable OneToOne relations

### DISCUSSION THREAD

**Discussion by JeongJuhyeon** [mikro-orm/mikro-orm#3578] [Q&A] [answered]
Suppose we have a User who may or may not own (at most one) Store.

```
@Entity()
export class User extends CustomBaseEntity {
  @Property()
  name!: string;

  @OneToOne(() => Store, (store) => store.owner, { owner: true })
  // A: store!: IdentifiedReference<Store, 'id'> | null;
  // B: store!: IdentifiedReference<Store | null, 'id'>;
  // C: store?: IdentifiedReference<Store, 'id'>;
}
```
Do any of these versions properly represent the actual typing? Simply ignoring the nullability would mean that after a `.load()` I'd get a non-optional Store which wouldn't be type-safe. 
If version C is correct, what should I be using instead of `.set()` when persisting a new Store on an existing User, given that apparently in that case `user.store` will be undefined?

**Comment by QuestionAndAnswer** [comment]
@JeongJuhyeon did you find an answer?

**Comment by B4nan** [comment, marked as answer]
Either A or C (or even their combination), both will work and are kinda correct, depends on what you want from the optional values, personally, I prefer not to use the `null` type explicitly. You can either use `set` to change the wrapped value, or preferably just replace the reference, as since v5.5 the referenes also hold the identity just like entities, so it is more correct to replace the whole reference wrapper (but still both should work fine).

B is wrong, you can't have an empty reference.

Note that the issue with `null` vs `undefined` is a bit tricky - a partially loaded entity might end up with `undefined` in place of any value (so not just nullable ones). Apart from that, you will get a `null` only for properties hydrated from database - you can force the value to be `undefined` via [forceUndefined](https://mikro-orm.io/docs/configuration#mapping-null-values-to-undefined).  New entity instance will have an `undefined` there unless you use a property initialized, e.g. `store!: Ref<Store> | null = null`.

**Reply by JeongJuhyeon** [reply]
Thanks, your suggestion (most closely corresponding to C) is what I ended up using. As for persisting, I just set store.user to the reference in the Store entity constructor.
